pennylane/templates/subroutines/iqp.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2018-2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the IQP template.
+"""
+from collections import defaultdict
+
+import numpy as np
+
+from pennylane.decomposition import (
+    add_decomps,
+    register_resources,
+    resource_rep,
+)
+from pennylane.operation import Operation
+from pennylane.ops import Hadamard, MultiRZ, PauliRot
+
+
+class IQP(Operation):
+    """
+    A template that builds an Instantaneous Quantum Polynomial (IQP) circuit.
+
+    In this `IQPOpt <https://arxiv.org/pdf/2501.04776>`__ paper and this
+    `train on classical, deploy on quantum <https://arxiv.org/pdf/2503.02934>`__ paper, the authors
+    present methods for analytically approximating expectation values coming from measurements made on IQP circuits.
+    This allows for the classical training of the parameters of these circuits prior to deploying them to a
+    quantum computer for actual computation.
+
+    Certain computational problems such as generative machine learning and combinatorial optimization can be cast as
+    a minimization over functions of these expectation values. Since these circuits are also expected to be hard to
+    sample from using classical algorithms, they can potentially lead to a quantum advantage.
+    """
+
+    resource_keys = {"spin_sym", "pattern", "num_wires"}
+
+    def __init__(
+        self, weights, num_wires, pattern, spin_sym=None, id=None
+    ):  # pylint: disable=too-many-arguments
+        r"""
+        IQP template corresponding to a parameterized IQP circuit. Based on `IQPopt: Fast optimization of
+        instantaneous quantum polynomial circuits in JAX <https://arxiv.org/pdf/2501.04776>`__.
+
+        Args:
+            weights (list): The parameters of the IQP gates.
+            num_wires (int): Number of wires in the circuit.
+            pattern (list[list[list[int]]]): Specification of the trainable gates. Each element of pattern corresponds to a
```

Comment:
Wouldn't `pattern` be `list[list[int]]`? Why is it nested three levels?

Reply:
Ok, so each sublist in `pattern` is a list of gates with the same parameter in `weights`, and each sublist in the sublist correspond to the wires that each gate acts on? Can you add an example in the docs for how it looks like?